File: test-project/src/Web/WebSocket.js
// WebSocket Client FFI for browser
// This module provides Effect-based bindings to the browser WebSocket API

export const create = (url) => () => {
  return new WebSocket(url);
};

export const createWithProtocolImpl = (url, protocol) => () => {
  return new WebSocket(url, protocol);
};

// State
export const readyStateImpl = (ws) => () => ws.readyState;
export const getUrl = (ws) => () => ws.url;
export const getProtocol = (ws) => () => ws.protocol;

// Sending
export const sendText = (ws) => (msg) => () => {
  ws.send(msg);
};

// Event handlers
export const onOpen = (ws) => (handler) => () => {
  ws.onopen = () => handler();
};

export const onCloseImpl = (ws, handler) => () => {
  ws.onclose = (event) => {
    const closeEvent = {
      code: event.code,
      reason: event.reason,
      wasClean: event.wasClean
    };
    handler(closeEvent)();
  };
};

export const onMessageImpl = (ws, handler) => () => {
  ws.onmessage = (event) => handler(event)();
};

export const onErrorImpl = (ws, handler) => () => {
  ws.onerror = () => handler();
};

// Lifecycle
export const close = (ws) => () => {
  ws.close();
};

export const closeWithCodeImpl = (ws, code, reason) => () => {
  ws.close(code, reason);
};

// Event data extraction
export const getMessageData = (event) => {
  const data = event.data;
  if (typeof data === "string") {
    return data;
  } else if (data instanceof ArrayBuffer) {
    const decoder = new TextDecoder();
    return decoder.decode(data);
  }
  return String(data);
};

File: test-project/src/Web/WebSocket.py
# FFI stub for PureScript compiler - actual implementation is in WebSocket.js
# This module is designed for browser JavaScript only.

def browser_only(name):
    raise Exception(
        f"Web.WebSocket.{name}: This module is for browser JavaScript only. "
        f"Use Server.WebSocket for Python WebSocket server operations."
    )

create = lambda url: lambda: browser_only("create")
createWithProtocolImpl = lambda url, protocol: lambda: browser_only("createWithProtocol")
readyStateImpl = lambda ws: lambda: browser_only("readyState")
getUrl = lambda ws: lambda: browser_only("getUrl")
getProtocol = lambda ws: lambda: browser_only("getProtocol")
sendText = lambda ws: lambda msg: lambda: browser_only("sendText")
onOpen = lambda ws: lambda handler: lambda: browser_only("onOpen")
onCloseImpl = lambda ws, handler: lambda: browser_only("onClose")
onMessageImpl = lambda ws, handler: lambda: browser_only("onMessage")
onErrorImpl = lambda ws, handler: lambda: browser_only("onError")
close = lambda ws: lambda: browser_only("close")
closeWithCodeImpl = lambda ws, code, reason: lambda: browser_only("closeWithCode")
getMessageData = lambda event: browser_only("getMessageData")
